master/buildbot/reporters/utils.py

```diff
@@ -0,0 +1,156 @@
+        prev = yield master.data.get(("builders", build['builderid'], "builds", n))
+        if prev['results'] != RETRY:
+            defer.returnValue(prev)
+        n -= 1
+    defer.returnValue(None)
+
+
+@defer.inlineCallbacks
+def getDetailsForBuildset(master, bsid, wantProperties=False, wantSteps=False,
+                          wantPreviousBuild=False, wantLogs=False):
+    # Here we will do a bunch of data api calls on behalf of the reporters
+    # We do try to make *some* calls in parallel with the help of gatherResults, but don't commit
+    # to much in that. The idea is to do parallelism while keeping the code readable
+    # and maintainable.
+
+    # first, just get the buildset and all build requests for our buildset id
+    dl = [master.data.get(("buildsets", bsid)),
+          master.data.get(('buildrequests', ),
+                          filters=[resultspec.Filter('buildsetid', 'eq', [bsid])])]
+    (buildset, breqs) = yield defer.gatherResults(dl)
+    # next, get the bdictlist for each build request
+    dl = [master.data.get(("buildrequests", breq['buildrequestid'], 'builds'))
+          for breq in breqs]
+
+    builds = yield defer.gatherResults(dl)
+    builds = flatten(builds, types=(list, UserList))
+    if builds:
+        yield getDetailsForBuilds(master, buildset, builds, wantProperties=wantProperties,
+                                  wantSteps=wantSteps, wantPreviousBuild=wantPreviousBuild, wantLogs=wantLogs)
+
+    defer.returnValue(dict(buildset=buildset, builds=builds))
+
+
+@defer.inlineCallbacks
+def getDetailsForBuilds(master, buildset, builds, wantProperties=False, wantSteps=False,
+                        wantPreviousBuild=False, wantLogs=False):
+    builderids = set([build['builderid'] for build in builds])
+
+    builders = yield defer.gatherResults([master.data.get(("builders", _id))
+                                          for _id in builderids])
+
+    buildersbyid = dict([(builder['builderid'], builder) for builder in builders])
+
+    if wantProperties:
+        buildproperties = yield defer.gatherResults(
+            [master.data.get(("builds", build['buildid'], 'properties'))
+             for build in builds])
+    else:
+        buildproperties = range(len(builds))
+
+    if wantPreviousBuild:
+        prev_builds = yield defer.gatherResults(
+            [getPreviousBuild(master, build) for build in builds])
+    else:
+        prev_builds = range(len(builds))
+
+    if wantSteps:
+        buildsteps = yield defer.gatherResults(
+            [master.data.get(("builds", build['buildid'], 'steps'))
+             for build in builds])
+        if wantLogs:
+            for s in flatten(buildsteps, types=(list, UserList)):
+                s['logs'] = yield master.data.get(("steps", s['stepid'], 'logs'))
+                for l in s['logs']:
+                    l['content'] = yield master.data.get(("logs", l['logid'], 'contents'))
+
+    else:
+        buildsteps = range(len(builds))
```

Comment:
If we dont want the step, we still need a list of size len(build) in order to simplify the zip later